lib/Dialect/TritonGPU/Transforms/CombineTensorSelectAndIf.cpp

```diff
@@ -0,0 +1,127 @@
+#include "mlir/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+
+#include <memory>
+
+using namespace mlir;
+using namespace triton;
+using namespace triton::gpu;
+
+#define GEN_PASS_CLASSES
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
+
+class CombineTensorSelectAndIfPass
+    : public TritonGPUCombineTensorSelectAndIfBase<
+          CombineTensorSelectAndIfPass> {
+public:
+  explicit CombineTensorSelectAndIfPass() = default;
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    ModuleOp m = getOperation();
+
+    // Go over the arith.select ops, look if there is an if
+    // with the same condition.
+    SmallVector<std::pair<arith::SelectOp, scf::IfOp>, 4> selectIfCandidates;
+    m.walk([&](arith::SelectOp selectOp) {
+      // Look if there is an if in the same block, with the same condition.
+      auto *parentBlock = selectOp->getBlock();
+      Value condition = selectOp.getOperand(0);
+      // Get condition's users
+      for (auto user : condition.getUsers()) {
+        if (auto ifOp = dyn_cast<scf::IfOp>(user)) {
+          if (ifOp->getBlock() == parentBlock) {
+            selectIfCandidates.emplace_back(selectOp, ifOp);
+            continue;
+          }
+        }
+      }
+    });
+
+    llvm::MapVector<arith::SelectOp, scf::IfOp> selectToIf;
+
+    // Go over the candidates and check if the select and if can be combined.
+    for (auto [selectOp, ifOp] : selectIfCandidates) {
+      if (selectToIf.count(selectOp) > 0) {
+        continue;
+      }
+      // If needs to be dominated by the select. Since the select is in the
+      // same block as the if, we can just check the order of the operations.
+      if (!selectOp->isBeforeInBlock(ifOp)) {
+        continue;
+      }
+      // If needs to dominate all the select's uses.
+      bool allUsesDominated = true;
+      for (auto user : selectOp.getResult().getUsers()) {
+        if (ifOp->isAncestor(user)) {
+          // Select used by something in the `if` itself.
+          allUsesDominated = false;
+          break;
+        }
+        if (user->getBlock() != selectOp->getBlock()) {
+          // Used outside the block, but not in `if`, means it's dominated.
```

Comment:
I don't think that's true with regions as there could be an op in between the select and if with a nested region?

Reply:
Right, thanks for catching that!